float64 desiredX
float64 desiredY
float64 desiredZ
float64 desiredW
